ktor-io/common/src/io/ktor/utils/io/ByteChannel.kt

```diff
@@ -133,65 +135,99 @@ public class ByteChannel(public val autoFlush: Boolean = false) : ByteReadChanne
 
     override fun toString(): String = "ByteChannel[${hashCode()}]"
 
-    // Awaiting Slot
-    private val suspensionSlot: AtomicRef<Continuation<Unit>?> = atomic(null)
-
     private suspend fun sleepForRead(min: Int) {
         while (flushBufferSize + _readBuffer.size < min && _closedCause.value == null) {
             suspendCancellableCoroutine {
-//                readContinuation.init(it.intercepted())
-                trySuspendSlot(it) { flushBufferSize + _readBuffer.size < min && _closedCause.value == null }
-//                readContinuation.getOrThrow()
+                trySuspendSlot(Slot.Read(it)) { flushBufferSize + _readBuffer.size < min && _closedCause.value == null }
             }
         }
     }
 
     @OptIn(InternalAPI::class)
     private suspend fun sleepForWrite() {
-        while (flushBufferSize >= CHANNEL_MAX_SIZE) {
+        while (flushBufferSize >= CHANNEL_MAX_SIZE && _closedCause.value == null) {
             suspendCancellableCoroutine {
-//                writeContinuation.init(it.intercepted())
-                trySuspendSlot(it) { flushBufferSize >= CHANNEL_MAX_SIZE }
-//                writeContinuation.getOrThrow()
+                trySuspendSlot(Slot.Write(it)) { flushBufferSize >= CHANNEL_MAX_SIZE && _closedCause.value == null }
```

Comment:
if slots show up in the profile we can reuse them within the call

Reply:
There's just a small increase in continuation allocations when I run the benchmark test.  There should be less... I'll see if I can bring it down...